rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -901,69 +921,129 @@ void PlayerImpl::load_storage_content()
-        message_ptr = peek_next_message_from_queue();
-        continue;
+
+      std::lock_guard<std::mutex> lk_skip_message(skip_message_in_main_play_loop_mutex_);
+      if (rclcpp::ok()) {
+        // This means that the message we took from the queue's top was invalidated after a seek(),
+        // so we need to take a fresh element from the top of the queue, unless we have to stop
+        if (skip_message_in_main_play_loop_) {
+          skip_message_in_main_play_loop_ = false;
+          cancel_wait_for_next_message_ = false;
+          if (stop_playback_.load()) {
+            break;
+          } else {
+            message_ptr = take_next_message_from_queue();
+            continue;
+          }
+        }
+
+        const bool message_published = publish_message(message_ptr);
+        // If we tried to publish because of play_next(), jump the clock
+        if (play_next_.load()) {
+          clock_->jump(message_ptr->recv_timestamp);
+          // If we successfully played next, notify that we're done, otherwise keep trying
+          if (message_published) {
+            play_next_ = false;
+
+            std::lock_guard<std::mutex> lk(finished_play_next_mutex_);
+            finished_play_next_ = true;
+            play_next_result_ = true;
+            finished_play_next_cv_.notify_all();
+          }
+        }
       }
-      publish_message(message_ptr);
+      message_ptr = take_next_message_from_queue();
+    }
+
+    // At this point, we're at the end of the playback round, there are no more messages to play
+    // If we're still trying to play next or just got a request, we did not succeed
+    if (play_next_.exchange(false)) {
+      std::lock_guard<std::mutex> lk(finished_play_next_mutex_);
+      finished_play_next_ = true;
+      play_next_result_ = false;
+      finished_play_next_cv_.notify_all();
+    }
+
+    // While we're in a pause state, make sure we don't return if we happen to be at the end of the
+    // queue or playback round. However, if we get a request for play next during sleep_until(...),
+    // we need to stop waiting here and proceed to handle the play next request by doing another
+    // loop of this while().
+    while (!stop_playback_ && is_paused() && !play_next_.load() && rclcpp::ok()) {
+      clock_->sleep_until(clock_->now());
+    }
+    // If we ran out of messages and are not in pause state, it means we're done playing,
+    // unless play_next() is resuming and pausing the clock in order to wake us up
+    if (!is_paused() && !play_next_.load()) {
+      break;
+    }
+
+    // If we had run out of messages before but are starting to play next again, e.g., after a
+    // seek(), we need to take
+    if (play_next_.load()) {
+      std::lock_guard<std::mutex> lk(skip_message_in_main_play_loop_mutex_);
+      skip_message_in_main_play_loop_ = false;
+      cancel_wait_for_next_message_ = false;
+      message_ptr = take_next_message_from_queue();
```

Comment:
To make sure that we are in `play_next` after the `seek` need to add a check under the mutex
`if (skip_message_in_main_play_loop_) {..}`
But in this case, it seems will be exact the same logic as in the lines: 984-997. It seems it will be the same logic if we would remove this block with `if (play_next_.load())` and just let the loop continue.

Reply:
sorry, I don't understand your comment. `play_next_` cannot turn to `false` after being `true` if `play_messages_from_queue()` doesn't notify `play_next()`.

Reply:
Okay, we agreed that we don;t need if